### PR TITLE
Introduce Pagination

### PR DESCRIPTION
**Pull request by LionOps**
Fixes https://github.com/yola/pixabayjs/issues/7

Adding pagination to pixabayjs while also introducing mockagent. WIP as I haven't updated the README and I'd like some eyes on the actual code.

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -57,30 +88,47 @@ describe('pixabayjs', function() {
       client = Object.create(pixabay);
       client.authenticate(username, key);
 
-      client
-        .request()
-        .query(query)
+      request = client
+        .request();
```
Comment (kahnvex) [inline test.js]: `request = client.request();`?

File: test.js [hunk 2]
```diff
@@ -50,6 +54,33 @@ describe('pixabayjs', function() {
     var query;
 
     before(function(done) {
+      mockagent.target(superagent);
+
+      mockagent.url(pixabayUrl, function(res) {
```
Comment (kahnvex) [inline test.js]: `mockagent.get` would work and not patch as many http methods.

File: test.js [hunk 3]
```diff
@@ -50,6 +54,33 @@ describe('pixabayjs', function() {
...
+        var query = url.parse(this.url).query;
+        var queryParams = querystring.parse(query);
```
Comment (kahnvex) [inline test.js]: You can just do:

``` javascript
var queryParams = url.parse(this.url, true).query;
```

File: test.js [hunk 4]
```diff
@@ -50,6 +54,33 @@ describe('pixabayjs', function() {
...
+
+        var firstPage = queryParams.page !== '2';
```
Comment (kahnvex) [inline test.js]: Can we dumb down this logic a little. It is a little hard to parse. Would this work:

``` javascript
if (queryParams.page === 1) {
  responseBody = [0, 1, 2, 3];
} else if (queryParams.page === 2) {
  responseBody = [4, 5];
} else {
  throw new Error('Mock encountered an unexpected condition');
}
```

**Comment by kahnvex** [comment]
@DevShep we need to finalize this ASAP. A significant amount of this weeks work depends on it.

File: README.md [hunk 5]
```diff
@@ -26,15 +26,27 @@ pixabay.authenticate('username', 'api_key');
...
+// Create a request handler
+var handler = pixabay
     .request()
     .query({order: 'latest'})
-    .search(['dogs', 'puppies'])
-    .get();
+    .search(['dogs', 'puppies']);
+
+// Get the request promise
+var request = handler.get();
```
Comment (kahnvex) [inline README.md]: This usage example seems a little clunky. Why not do:

``` javascript
var requestPromise = pixabay.request()
  .query({order: 'latest'})
  .search(['dogs', 'puppies'])
  .get();  // Returns the request promise
```

This way we can completely abstract away the concept of a "request handler".

File: src/handler.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  return response;
+};
+
+Handler.prototype.next = function() {
+  if (!this._initRequest) {
+    throw new Error('No initial request has been made');
+  }
+
+  if (this.lastPage) {
+    throw new Error('Last page has been retrieved');
```
Comment (kahnvex) [inline src/handler.js]: Return a promise for `null`, this will be consistent with other APIs we've written. Also, in cases where you do need to throw an error, and you in a promise returning function, it is more appropriate to return a promise in the failure state, than synchronously throwing an error.

File: src/handler.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  this._totalPages = Math.ceil(this.totalHits / per_page);
+
+  if (this._page === this._totalPages) {
+    this.lastPage = true;
+  }
+
+  return response;
+};
+
+Handler.prototype.next = function() {
```
Comment (kahnvex) [inline src/handler.js]: I think it is odd to have the `Handler` also be a generator. Usually a generator is a distinct object/interface.

File: src/handler.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+var uniq = require('lodash.uniq');
+
+function Handler(retriever) {
```
Comment (kahnvex) [inline src/handler.js]: Can we call this something a little more specific than `Handler`?

File: README.md (hunk 5 above)
Comment (LionOps) [inline README.md]: There's a problem with that. `get` returns a promise of the request, not the handler, and the handler is needed to call `next`.

I'm open to ideas on how to better design it.

File: src/handler.js (hunk 8 above)
Comment (LionOps) [inline src/handler.js]: Ideas?

File: src/handler.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  return this.query({page: this._page});
+};
+
+Handler.prototype.query = function(obj) {
+  this._retriever.query(obj);
+  return this;
+};
+
+Handler.prototype.search = function(search) {
+  if (this._requested) {
```
Comment (kahnvex) [inline src/handler.js]: Will this ever be true?

File: src/handler.js (hunk 7 above)
Comment (LionOps) [inline src/handler.js]: How would you approach it? The problem I'm running into is that the responses don't give any information as to what the request is. They only give `totalHits` and `hits`, which is defaulted to a 20-element array of objects. The only difference between two pages is the hits array, but there'd be no way to determine which page is the first without some sort of meta knowledge.

File: src/handler.js (hunk 8 above)
Comment (kahnvex) [inline src/handler.js]: I think if we address the concern I raised [here](https://github.com/yola/pixabayjs/pull/8#discussion_r33066241) it will be easier to address this issue.

File: src/handler.js (hunk 7 above)
Comment (kahnvex) [inline src/handler.js]: > there'd be no way to determine which page is the first without some sort of meta knowledge.

I don't follow.

Comment (LionOps) [inline src/handler.js]: The response for page 1 looks like this:

```
{
  "totalHits": 139,
  "hits": [
    {
       ...
    },
    ...
  ]
}
```

The response for page 2 looks like this:

```
{
  "totalHits": 139,
  "hits": [
    {
       ...
    },
    ...
  ]
}
```

Comment (LionOps) [inline src/handler.js]: Would it make sense to remove `get`? It might make more sense to treat it as a generator than not. The methods would be used to set up the query params and then calling `next` would return a promise for the initial request. Calling it again would return a promise for the next page, etc.

File: src/list-factory.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+var ResponseHandler = require('./request-handler');
+
+function ListFactory(retriever, spec) {
```
Comment (euwest) [inline src/list-factory.js]: why `spec` here and `options` elsewhere? is there a difference you're trying to highlight?

File: README.md (hunk 5 above)
Comment (euwest) [inline README.md]: `request()` should be `requestFactory()` on line 31, yeah?

**Comment by LionOps** [comment]
Working on the Readme now

**Comment by LionOps** [comment]
:bell: Ready for review.

**Comment by kahnvex** [comment]
Can we expose a `.get()` function from the request factory, as well as exposing a result list factory? It should be a simple delegating call to retriever's get function.

File: src/result-list.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  this._retriever.query({page: this._page});
+  this._page += 1;
+  return this._get(this._page - 1);
+};
+
+ResultList.prototype._get = function(page) {
+  return this._retriever
+    .get()
+    .then(bind(this._handleSuccess, this, '_', page))
+    .fail(bind(this._handleFailure, this, '_', page));
```
Comment (kahnvex) [inline src/result-list.js]: Just switch the order of the arguments in `_handleSuccess` and `_handleFailure` and avoid the confusion of placeholders.

**Comment by kahnvex** [comment]
Configure @snitch to lint this repo?

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

**Comment by snitch** [comment]
:sparkles: No lint errors found. :sparkles:

File: src/request-factory.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+
+var ResultList = require('./result-list');
+
+function RequestFactory(retriever) {
+  this._retriever = retriever;
+}
+
+RequestFactory.prototype.get = function(options) {
+  return this.resultList(options).next();
```
Comment (kahnvex) [inline src/request-factory.js:10]: This is not what I meant. I meant that this should delegate to the retriever. Delegating to the resultList doesn't make any sense. We don't even know if it is defined.

Comment (kahnvex) [inline src/request-factory.js:10]: I take that back, it will be defined. It is still a little weird, but this will work.

**Comment by kahnvex** [comment]
:+1:

File: README.md [hunk 13]
```diff
@@ -1,5 +1,5 @@
 # pixabayjs
-A Javascript Wrapper for the Pixabay API. To learn more about Pixabay, read the [docs](http://pixabay.com/api/docs/).
+A Javascript Wrapper for the Pixabay API. To learn more about Pixabay, read the [docs][docs].
```
Comment (stefanor) [inline README.md:2]: Simple syntax: `[docs][]`
